evaluator/evaluator.py

```diff
@@ -113,6 +113,51 @@ def _store_cves(self, system_id, system_cves, source_name="VMAAS"):
                                                                                       tuple(mitigated_cves),))
             cur.close()
             self.conn.commit()
+
+    def _sync_cve_md(self, page_size=5000, page=1, cves_in_db=None):
+        cur = self.conn.cursor()
+        if not cves_in_db:
+            cur.execute('select cve from cve_metadata')
+            cves_in_db = []
+            for cve_tuple in cur.fetchall():
+                cves_in_db.append(cve_tuple[0])
+        cve_list = {'cve_list' : [".*"], 'page_size' : page_size, 'page': page}
+        response = requests.post(self.vmaas_cves_endpoint, json=cve_list)
+        if response.status_code != 200:
+            return
+        r_json = response.json()
+        logging.info('syncing cve_metadata from vmaas: page: %s, page_size: %s, pages: %s' % (page, page_size, r_json['pages']))
+        cves = r_json['cve_list']
+        to_insert = []
+        to_update = []
+        for cve in cves:
+            description = cves[cve]['description']
+            impact = cves[cve]['impact']
+            public_date = cves[cve]['public_date'] or None
+            modified_date = cves[cve]['modified_date'] or None
+            cvss3_score = float(cves[cve]['cvss3_score']) if cves[cve]['cvss3_score'] else None
+            cvss3_metrics = cves[cve]['cvss3_metrics']
+            row = (cve, description, impact, public_date, modified_date, cvss3_score, cvss3_metrics)
+            if cve not in cves_in_db:
+                to_insert.append(row)
+            else:
+                to_update.append(row)
+        if to_insert:
+            execute_values(cur, """insert into cve_metadata
+                                (cve, description, impact, public_date, modified_date, cvss3_score, cvss3_metrics)
+                                values %s""", to_insert, page_size=len(to_insert))
+        if to_update:
+            execute_values(cur, """update cve_metadata set description = data.description, impact = data.impact,
+                                public_date = cast(data.public_date as timestamp with time zone),
+                                modified_date = cast(data.modified_date as timestamp with time zone),
+                                cvss3_score = data.cvss3_score, cvss3_metrics = data.cvss3_metrics
+                                from (values %s) as data (cve, description, impact, public_date, modified_date,
+                                cvss3_score, cvss3_metrics) where cve_metadata.cve = data.cve""",
+                           to_update, page_size=len(to_update))
+        cur.close()
+        self.conn.commit()
+        if page < r_json['pages']:
+            self._sync_cve_md(page_size, page + 1, cves_in_db)
```

Comment:
line 160:

1) Looks like there should be `cves_in_db + to_insert` (or just names from to_insert) because we inserted new values into DB already.
2) The minimum `page_size` should be around 200. Otherwise, there should be an exception.
```python
>>> print(sys.getrecursionlimit())
1000
```

Maybe in this case just make a loop 
```python
while page <= r_json['pages']:
    # do request, store into DB
    page = page +1

Reply:
1. No, these two variables are not compatible with themselves, `cves_in_db` is a list o str, while `to_insert` is list of tuples. Additionally, there's no need to do so, given how the algorithm works as there are no duplicate cve records returned by vmaas on next pages for the api call we're using.
Doing `cves_in_db + to_insert` would just slow the whole import (additionally, if the situation that would require this construction by any chance occurred, it would be discovered by database constraints).

2. Vmaas returns by default 5000 records per page so I've just took default page_size from vmaas. I don't see what's wrong with that. In that case we should also change default page_size in vmaas.

As for the recursion - it's much more elegant solution as we need to know `r_json['pages'] before the loop. Your suggestion would require one call to vmaas before the actual import loop just to get number of pages.

Reply:
1) OK.
2) Sorry, maybe my comment wasn't so clear. I mean 5000 is OK, but this value is passed as a parameter and if it is too small we can get `RuntimeError: maximum recursion depth exceeded`